subsys/usb/device_next/class/bt_hci.c

```diff
@@ -0,0 +1,491 @@
+		      const struct net_buf *const buf)
+{
+	struct net_buf *cmd_buf;
+
+	/* We expect host-to-device class request */
+	if (setup->RequestType.type != USB_REQTYPE_TYPE_CLASS) {
+		errno = -ENOTSUP;
+		return 0;
+	}
+
+	LOG_DBG("bmRequestType 0x%02x bRequest 0x%02x",
+		setup->bmRequestType, setup->bRequest);
+
+	cmd_buf = bt_buf_get_tx(BT_BUF_CMD, K_NO_WAIT, buf->data, buf->len);
+	if (!cmd_buf) {
+		LOG_ERR("Cannot get free buffer");
+		return -ENOMEM;
+	}
+
+	net_buf_put(&bt_hci_rx_queue, cmd_buf);
+
+	return 0;
+}
+
+static int bt_hci_init(struct usbd_class_node *const c_nd)
+{
+	return 0;
+}
+
+static struct usbd_class_api bt_hci_api = {
+	.request = bt_hci_request,
+	.update = bt_hci_update,
+	.enable = bt_hci_enable,
+	.disable = bt_hci_disable,
+	.control_to_dev = bt_hci_ctd,
+	.init = bt_hci_init,
+};
+
+#define BT_HCI_DESCRIPTOR_DEFINE(n)						\
+static struct usbd_bt_hci_desc bt_hci_desc_##n = {				\
+	.if0 = {								\
+		.bLength = sizeof(struct usb_if_descriptor),			\
+		.bDescriptorType = USB_DESC_INTERFACE,				\
+		.bInterfaceNumber = 0,						\
+		.bAlternateSetting = 0,						\
+		.bNumEndpoints = 3,						\
+		.bInterfaceClass = USB_BCC_WIRELESS_CONTROLLER,			\
+		.bInterfaceSubClass = USB_BT_HCI_SUBCLASS,			\
+		.bInterfaceProtocol = USB_BT_HCI_PROTOCOL,			\
+		.iInterface = 0,						\
+	},									\
+										\
+	.if0_int_ep = {								\
+		.bLength = sizeof(struct usb_ep_descriptor),			\
+		.bDescriptorType = USB_DESC_ENDPOINT,				\
+		.bEndpointAddress = 0x81,					\
+		.bmAttributes = USB_EP_TYPE_INTERRUPT,				\
+		.wMaxPacketSize = sys_cpu_to_le16(BT_HCI_DEFAULT_INT_EP_MPS),	\
+		.bInterval = 1,							\
+	},									\
+										\
+	.if0_out_ep = {								\
+		.bLength = sizeof(struct usb_ep_descriptor),			\
+		.bDescriptorType = USB_DESC_ENDPOINT,				\
+		.bEndpointAddress = 0x02,					\
+		.bmAttributes = USB_EP_TYPE_BULK,				\
+		.wMaxPacketSize = sys_cpu_to_le16(BT_HCI_DEFAULT_BULK_EP_MPS),	\
+		.bInterval = 0,							\
+	},									\
+										\
+	.if0_in_ep = {								\
+		.bLength = sizeof(struct usb_ep_descriptor),			\
+		.bDescriptorType = USB_DESC_ENDPOINT,				\
+		.bEndpointAddress = 0x82,					\
+		.bmAttributes = USB_EP_TYPE_BULK,				\
+		.wMaxPacketSize = sys_cpu_to_le16(BT_HCI_DEFAULT_BULK_EP_MPS),	\
+		.bInterval = 0,							\
+	},									\
+};
+
+#define BT_HCI_CLASS_DATA_DEFINE(n)						\
+	static struct bt_hci_data bt_hci_data_##n = {				\
+		.sync_sem = Z_SEM_INITIALIZER(bt_hci_data_##n.sync_sem, 0, 1),	\
+	};									\
+										\
+	static struct usbd_class_data bt_hci_class_##n = {			\
+		.desc = (struct usb_desc_header *)&bt_hci_desc_##n,		\
+		.priv = &bt_hci_data_##n,					\
+		.v_reqs = &bt_hci_vregs,					\
+	};									\
+										\
+	USBD_DEFINE_CLASS(bt_hci_##n, &bt_hci_api, &bt_hci_class_##n);
+
+/*
+ * Bluetooth subsystem does not support multiple HCI instances,
+ * but we are almost ready for it.
+ */
+BT_HCI_DESCRIPTOR_DEFINE(0)
+BT_HCI_CLASS_DATA_DEFINE(0)
```

Comment:
I think this relates to #51503 too. Along with making the HCI driver DTS-based, it becomes more natural to declare multiple instances. That said, the host stack treats the controller as singleton. There's a single `bt_dev` struct that's accessed as a global variable in several places of the host. To make the host truly multi-HCI aware there would need to be this `bt_dev` context per controller instance that's then passed around as a parameter in most existing Bluetooth APIs.

Reply:
@jhedberg @jfischer-no does anything need to be done here?

Reply:
As I understood @carlescufi, not in the near future.